src/QsCompiler/Tests.Compiler/LambdaLiftingTests.fs

```diff
@@ -0,0 +1,573 @@
+
+    [<Fact>]
+    [<Trait("Category", "Return Values")>]
+    member this.``Call Unit Callable Recursive``() =
+        let result = CompileLambdaLiftingTest 6
+
+        let original = GetCallableWithName result Signatures.LambdaLiftingNS "Foo" |> GetBodyFromCallable
+
+        let generated =
+            GetCallablesWithSuffix result Signatures.LambdaLiftingNS "_Foo"
+            |> (fun x ->
+                Assert.True(1 = Seq.length x)
+                Seq.item 0 x |> GetBodyFromCallable)
+
+        let lines = generated |> GetLinesFromSpecialization
+
+        Assert.True(
+            1 = Seq.length lines,
+            sprintf "Callable %O(%A) did not have the expected number of statements." generated.Parent generated.Kind
+        )
+
+        Assert.True(
+            lines.[0] = "Microsoft.Quantum.Testing.LambdaLifting.Foo();",
+            "The generated callable did not have the expected content."
+        )
+
+        let lines = original |> GetLinesFromSpecialization
+
+        Assert.True(
+            1 = Seq.length lines,
+            sprintf "Callable %O(%A) did not have the expected number of statements." original.Parent original.Kind
+        )
+
+        let expected = sprintf "let lambda = %O(_);" generated.Parent
+        Assert.True(lines.[0] = expected, "The generated call expression did not have the correct arguments.")
+
+    [<Fact>]
+    [<Trait("Category", "Parameters")>]
+    member this.``Use Closure``() =
+        let result = CompileLambdaLiftingTest 7
+
+        let original = GetCallableWithName result Signatures.LambdaLiftingNS "Foo" |> GetBodyFromCallable
+
+        let generated =
+            GetCallablesWithSuffix result Signatures.LambdaLiftingNS "_Foo"
+            |> (fun x ->
+                Assert.True(1 = Seq.length x)
+                Seq.item 0 x |> GetBodyFromCallable)
+
+        let lines = generated |> GetLinesFromSpecialization
+
+        Assert.True(
+            1 = Seq.length lines,
+            sprintf "Callable %O(%A) did not have the expected number of statements." generated.Parent generated.Kind
+        )
+
+        Assert.True(lines.[0] = "return (x, y, z);", "The generated callable did not have the expected content.")
+
+        let lines = original |> GetLinesFromSpecialization
+
+        Assert.True(
+            5 = Seq.length lines,
+            sprintf "Callable %O(%A) did not have the expected number of statements." original.Parent original.Kind
+        )
+
+        let expected = sprintf "let lambda = %O(x, y, z, _);" generated.Parent
+        Assert.True(lines.[4] = expected, "The generated call expression did not have the correct arguments.")
+
+    [<Fact>]
+    [<Trait("Category", "Parameters")>]
+    member this.``Use Lots of Params``() = CompileLambdaLiftingTest 8 |> ignore
+
+    [<Fact>]
+    [<Trait("Category", "Parameters")>]
+    member this.``Use Closure With Params``() = CompileLambdaLiftingTest 9 |> ignore
```

Comment:
That the result from compilation has the appropriate signatures is enough for this test case. That is handled in the signature check.

Reply:
I'd be interested in some additional assertions, like

- What are the names of the parameters of the lifted lambda?
- What's the body of the lifted lambdas? Does it preserve the names of the arguments and the enclosed variables?
- Does the original lambda location get replaced with the right form of partial application? Are the arguments/closure tuple in the right order?

Reply:
I don't think I can reasonable accomplish those assertions with these test case ("With Lots of Params" and "Use Closure With Params"). I've added some additional tests since this comment was posted, and I believe the second bullet point is a covered scenario now. If you ignore these two tests and look at the rest of the tests, what scenarios would you say still need coverage?

Reply:
I've updated the tests ("With Lots of Params" and "Use Closure With Params") to check the parameter names of the lambdas, that the correct lambdas are called in the right places, and that the argument tuples are as expected. I've added some additional tests since this comment was posted, and I believe the second bullet point is a covered scenario now as well, but by other tests since these tests are meant to focus on the parameters.